File: src/frontend/lexer/scanner.rs
//! Character-level scanner with PUA-aware UTF-8 handling.
//!
//! This module provides the lowest-level input abstraction for the BCC lexer.
//! The [`Scanner`] reads individual characters from the preprocessor's expanded
//! source string, provides lookahead buffering (at least 2 characters), tracks
//! position (byte offset, line number, column number), and handles PUA-encoded
//! bytes transparently.
//!
//! # PUA Transparency
//!
//! Non-UTF-8 bytes in C source files are encoded as Unicode Private Use Area
//! (PUA) code points (U+E080–U+E0FF) during source file reading. The scanner
//! treats these as normal characters — they flow through to string literals
//! and identifiers without special treatment. Since the source is a valid
//! Rust `&str`, all PUA code points are valid UTF-8-encoded Unicode characters
//! and [`CharIndices`] handles them correctly.
//!
//! # Line Ending Normalization
//!
//! The scanner normalizes all line endings to `\n`:
//! - `\n` → returned as `\n`
//! - `\r\n` → consumed as a unit, returned as `\n`
//! - `\r` alone → returned as `\n`
//!
//! Byte offsets correctly account for the actual bytes consumed (1 for `\n`
//! or standalone `\r`, 2 for `\r\n`).
//!
//! # Position Tracking
//!
//! Positions use byte offsets (0-indexed) for `Span` construction, compatible
//! with [`crate::common::diagnostics::Span`]. Line numbers are 1-indexed,
//! column numbers are 1-indexed and measured in bytes (not characters).

use crate::common::encoding;

use std::iter::Peekable;
use std::str::CharIndices;

// ---------------------------------------------------------------------------
// Position — source location tracking
// ---------------------------------------------------------------------------

/// Source position within a file — byte offset, line, and column.
///
/// The byte offset is the definitive position for `Span` construction. The
/// line and column numbers are for human-readable diagnostic display only.
///
/// # Invariants
///
/// - `offset` is a 0-indexed byte offset from the start of the source string.
/// - `line` starts at 1 (first line of the file).
/// - `column` starts at 1 (first byte of the line), measured in bytes not
///   characters.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct Position {
    /// Byte offset from the start of the source string (0-indexed).
    pub offset: u32,
    /// Line number (1-indexed).
    pub line: u32,
    /// Column number (1-indexed, in bytes not characters).
    pub column: u32,
}

impl Position {
    /// Create a new position with explicit values.
    #[inline]
    pub fn new(offset: u32, line: u32, column: u32) -> Self {
        Self { offset, line, column }
    }
}

// ---------------------------------------------------------------------------
// Scanner — character-level input with lookahead and position tracking
// ---------------------------------------------------------------------------

/// Character-level scanner for the BCC lexer.
///
/// Reads individual characters from a preprocessor-expanded source string,
/// provides bounded lookahead (via [`peek`], [`peek_nth`], and [`unget`]),
/// tracks byte offset / line / column, and normalizes line endings.
///
/// PUA-encoded code points (U+E080–U+E0FF) pass through transparently as
/// normal Unicode characters.
///
/// # Example
///
/// ```ignore
/// use bcc::frontend::lexer::scanner::Scanner;
///
/// let mut sc = Scanner::new("hello\nworld");
/// assert_eq!(sc.advance(), Some('h'));
/// assert_eq!(sc.peek(), Some('e'));
/// assert_eq!(sc.line(), 1);
/// ```
///
/// [`peek`]: Scanner::peek
/// [`peek_nth`]: Scanner::peek_nth
/// [`unget`]: Scanner::unget
pub struct Scanner<'src> {
    /// The full source string (PUA-encoded for non-UTF-8 files).
    source: &'src str,

    /// Peekable iterator over characters with their byte indices.
    chars: Peekable<CharIndices<'src>>,

    /// Current byte offset in the source string — position of the next
    /// character to be read (when the lookahead buffer is empty).
    offset: usize,

    /// Current line number (1-indexed).
    line: u32,

    /// Current column number (1-indexed, in bytes).
    column: u32,

    /// Lookahead buffer storing `(character, position_of_character)` pairs.
    ///
    /// Each entry's [`Position`] represents the scanner state immediately
    /// **before** consuming that character. Characters are pushed here by
    /// [`unget`] or pre-read by [`peek_nth`] when the lexer needs to look
    /// further ahead than what the underlying iterator provides.
    ///
    /// [`unget`]: Scanner::unget
    /// [`peek_nth`]: Scanner::peek_nth
    lookahead: Vec<(char, Position)>,
}

impl<'src> Scanner<'src> {
    // ===================================================================
    // Construction
    // ===================================================================

    /// Create a new scanner for the given source string.
    ///
    /// The scanner starts at byte offset 0, line 1, column 1.
    pub fn new(source: &'src str) -> Self {
        Scanner {
            source,
            chars: source.char_indices().peekable(),
            offset: 0,
            line: 1,
            column: 1,
            lookahead: Vec::with_capacity(4),
        }
    }

    // ===================================================================
    // Core character operations
    // ===================================================================

    /// Advance to the next character, consuming it.
    ///
    /// Returns `Some(char)` with the next character (after line-ending
    /// normalization), or `None` at end-of-file.
    ///
    /// # Line-Ending Normalization
    ///
    /// - `\n` is returned as `\n` (1 byte consumed).
    /// - `\r\n` is returned as `\n` (2 bytes consumed).
    /// - `\r` alone is returned as `\n` (1 byte consumed).
    ///
    /// All other characters, including PUA code points (U+E080–U+E0FF),
    /// are returned as-is.
    pub fn advance(&mut self) -> Option<char> {
        // Priority 1: consume from the lookahead buffer.
        if !self.lookahead.is_empty() {
            let (ch, pos) = self.lookahead.remove(0);
            // Determine the actual byte length this character spans in the
            // original source. For normalized newlines this may be 1 or 2.
            let byte_len = self.source_byte_len(ch, pos.offset as usize);
            // Update scanner state to the position AFTER this character.
            self.offset = pos.offset as usize + byte_len;
            if ch == '\n' {
                self.line = pos.line + 1;
                self.column = 1;
            } else {
                self.line = pos.line;
                self.column = pos.column + byte_len as u32;
            }
            return Some(ch);
        }

        // Priority 2: read directly from the character iterator.
        let (byte_idx, raw_ch) = self.chars.next()?;

        match raw_ch {
            '\r' => {
                // Normalize \r\n (Windows) or standalone \r (old Mac) to \n.
                if let Some(&(_, '\n')) = self.chars.peek() {
                    self.chars.next(); // consume the '\n' half of \r\n
                    self.offset = byte_idx + 2;
                } else {
                    self.offset = byte_idx + 1;
                }
                self.line += 1;
                self.column = 1;
                Some('\n') // normalized
            }
            '\n' => {
                self.offset = byte_idx + 1;
                self.line += 1;
                self.column = 1;
                Some('\n')
            }
            ch => {
                let len = ch.len_utf8();
                self.offset = byte_idx + len;
                self.column += len as u32;
                Some(ch)
            }
        }
    }

    /// Peek at the next character without consuming it.
    ///
    /// Returns `Some(char)` with the next character (after line-ending
    /// normalization), or `None` at end-of-file.
    ///
    /// This method does **not** advance the scanner's position.
    pub fn peek(&mut self) -> Option<char> {
        // Check the lookahead buffer first.
        if let Some(&(ch, _)) = self.lookahead.first() {
            return Some(ch);
        }
        // Peek at the raw iterator with \r normalization.
        self.chars.peek().map(|&(_, ch)| {
            if ch == '\r' { '\n' } else { ch }
        })
    }

    /// Look ahead `n` characters (0-indexed) without consuming.
    ///
    /// `peek_nth(0)` is equivalent to [`peek()`](Scanner::peek).
    /// `peek_nth(1)` returns the character after the next one, and so on.
    ///
    /// The lookahead buffer is filled as needed by pre-reading from the
    /// underlying iterator. All pre-read characters undergo line-ending
    /// normalization.
    pub fn peek_nth(&mut self, n: usize) -> Option<char> {
        // If already buffered, return immediately.
        if n < self.lookahead.len() {
            return Some(self.lookahead[n].0);
        }
        // Fill the lookahead up to n+1 entries.
        self.fill_lookahead(n + 1);
        self.lookahead.get(n).map(|&(ch, _)| ch)
    }

    /// Push a character back to the front of the scanner's input.
    ///
    /// The `pos` argument must be the [`Position`] that the scanner had
    /// **before** this character was consumed — i.e., the value returned by
    /// [`position()`](Scanner::position) immediately before the
    /// corresponding [`advance()`](Scanner::advance) call.
    ///
    /// # Usage
    ///
    /// Used when the lexer "over-reads" a character and needs to put it
    /// back for re-scanning.
    pub fn unget(&mut self, ch: char, pos: Position) {
        self.lookahead.insert(0, (ch, pos));
        // Restore the scanner's current-position state so that position(),
        // offset(), line(), column() all reflect this character's location
        // (the next character to be read).
        self.offset = pos.offset as usize;
        self.line = pos.line;
        self.column = pos.column;
    }

    // ===================================================================
    // Position query methods
    // ===================================================================

    /// Return the current scanner position.
    ///
    /// This is the position of the **next** character to be read — i.e.,
    /// what [`advance()`](Scanner::advance) would consume.
    #[inline]
    pub fn position(&self) -> Position {
        Position {
            offset: self.offset as u32,
            line: self.line,
            column: self.column,
        }
    }

    /// Return the current byte offset (shorthand for `position().offset`).
    #[inline]
    pub fn offset(&self) -> u32 {
        self.offset as u32
    }

    /// Return the current line number (1-indexed).
    #[inline]
    pub fn line(&self) -> u32 {
        self.line
    }

    /// Return the current column number (1-indexed, in bytes).
    #[inline]
    pub fn column(&self) -> u32 {
        self.column
    }

    // ===================================================================
    // Utility methods
    // ===================================================================

    /// Check whether the scanner has reached end-of-file.
    ///
    /// Returns `true` when there are no more characters to read.
    #[inline]
    pub fn is_eof(&mut self) -> bool {
        self.peek().is_none()
    }

    /// Return the full source string.
    ///
    /// Useful for extracting substrings between known byte offsets.
    #[inline]
    pub fn source(&self) -> &'src str {
        self.source
    }

    /// Return a substring of the source between byte offsets `start`
    /// (inclusive) and `end` (exclusive).
    ///
    /// Both offsets must lie on valid UTF-8 character boundaries within
    /// the source string.
    ///
    /// # Panics
    ///
    /// Panics if `start > end`, if either offset exceeds the source
    /// length, or if an offset is not on a UTF-8 boundary.
    #[inline]
    pub fn slice(&self, start: usize, end: usize) -> &'src str {
        &self.source[start..end]
    }

    /// Consume characters as long as `predicate` returns `true`.
    ///
    /// Stops at the first character for which `predicate` returns `false`,
    /// or at end-of-file.
    pub fn skip_while(&mut self, predicate: impl Fn(char) -> bool) {
        loop {
            match self.peek() {
                Some(ch) if predicate(ch) => {
                    self.advance();
                }
                _ => break,
            }
        }
    }

    /// If the next character equals `ch`, consume it and return `true`.
    /// Otherwise return `false` without consuming.
    ///
    /// Used for optional character matching (e.g., checking for `=`
    /// after `+` to form `+=`).
    pub fn consume_if(&mut self, ch: char) -> bool {
        if self.peek() == Some(ch) {
            self.advance();
            true
        } else {
            false
        }
    }

    /// If the next character satisfies `pred`, consume and return
    /// `Some(ch)`. Otherwise return `None` without consuming.
    pub fn consume_if_pred(&mut self, pred: impl Fn(char) -> bool) -> Option<char> {
        match self.peek() {
            Some(ch) if pred(ch) => {
                self.advance();
                Some(ch)
            }
            _ => None,
        }
    }

    // ===================================================================
    // PUA utility
    // ===================================================================

    /// Check if a character is a PUA-encoded non-UTF-8 byte (U+E080–U+E0FF).
    ///
    /// Convenience wrapper around [`encoding::is_pua_encoded`]. The scanner
    /// passes PUA code points through transparently, but callers (such as
    /// the lexer or diagnostic system) may use this to identify characters
    /// that represent original non-UTF-8 source bytes.
    ///
    /// [`encoding::is_pua_encoded`]: crate::common::encoding::is_pua_encoded
    #[inline]
    pub fn is_pua_char(ch: char) -> bool {
        encoding::is_pua_encoded(ch)
    }

    // ===================================================================
    // Internal helpers
    // ===================================================================

    /// Fill the lookahead buffer until it contains at least `target_len`
    /// entries.
    ///
    /// Pre-reads characters from the underlying [`CharIndices`] iterator,
    /// computes their positions, and applies line-ending normalization.
    /// Stops early if the iterator is exhausted (EOF).
    fn fill_lookahead(&mut self, target_len: usize) {
        while self.lookahead.len() < target_len {
            // Compute the position this next character will occupy.
            let pos = self.next_lookahead_position();

            // Read the next raw character from the iterator.
            match self.chars.next() {
                Some((_, '\r')) => {
                    // Normalize \r\n or standalone \r to \n.
                    if let Some(&(_, '\n')) = self.chars.peek() {
                        self.chars.next(); // consume the '\n' of \r\n
                    }
                    self.lookahead.push(('\n', pos));
                }
                Some((_, ch)) => {
                    self.lookahead.push((ch, pos));
                }
                None => break, // EOF reached
            }
        }
    }

    /// Compute the [`Position`] for the next character to be appended to
    /// the lookahead buffer.
    ///
    /// If the buffer is empty, this equals the current scanner position.
    /// Otherwise it is the position immediately after the last buffered
    /// character.
    fn next_lookahead_position(&self) -> Position {
        match self.lookahead.last() {
            Some(&(ch, pos)) => {
                let byte_len = self.source_byte_len(ch, pos.offset as usize);
                if ch == '\n' {
                    Position {
                        offset: pos.offset + byte_len as u32,
                        line: pos.line + 1,
                        column: 1,
                    }
                } else {
                    Position {
                        offset: pos.offset + byte_len as u32,
                        line: pos.line,
                        column: pos.column + byte_len as u32,
                    }
                }
            }
            None => Position {
                offset: self.offset as u32,
                line: self.line,
                column: self.column,
            },
        }
    }

    /// Determine how many bytes a character occupies at a given source
    /// offset.
    ///
    /// For most characters this is simply `ch.len_utf8()`. For normalised
    /// newlines (stored as `'\n'` in the lookahead) we inspect the raw
    /// source bytes to distinguish between `\n` (1 byte), standalone `\r`
    /// (1 byte), and `\r\n` (2 bytes).
    #[inline]
    fn source_byte_len(&self, ch: char, at_offset: usize) -> usize {
        if ch == '\n' {
            let bytes = self.source.as_bytes();
            match bytes.get(at_offset) {
                Some(&b'\r') => {
                    // \r followed by \n is 2 bytes; standalone \r is 1.
                    if bytes.get(at_offset + 1) == Some(&b'\n') {
                        2
                    } else {
                        1
                    }
                }
                Some(&b'\n') => 1,
                // Defensive fallback — should not be reached in normal
                // operation.
                _ => 1,
            }
        } else {
            ch.len_utf8()
        }
    }
}

File: gophc-bcc/frontend/lexer/scanner.go
package lexer

import (
	"github.com/gophc/blitzy-c-compiler/gophc-bcc/common"
)

type Position struct {
	Offset uint32
	Line   uint32
	Column uint32
}

func NewPosition(offset uint32, line uint32, column uint32) Position {
	return Position{
		Offset: offset,
		Line:   line,
		Column: column,
	}
}

type Scanner struct {
	source    string
	offset    int
	line      uint32
	column    uint32
	lookahead [][2]interface{}
}

func NewScanner(source string) *Scanner {
	return &Scanner{
		source:    source,
		offset:    0,
		line:      1,
		column:    1,
		lookahead: make([][2]interface{}, 0),
	}
}

func (s *Scanner) advance() rune {
	ch, _ := s.doAdvance()
	if ch == 0 {
		return 0
	}

	if ch == '\n' {
		s.line++
		s.column = 1
	} else {
		s.column++
	}
	return ch
}

func (s *Scanner) doAdvance() (rune, Position) {
	if len(s.lookahead) > 0 {
		item := s.lookahead[0]
		s.lookahead = s.lookahead[1:]
		ch := item[0].(rune)
		pos := item[1].(Position)

		if ch == '\n' {
			s.offset = int(pos.Offset) + 1
		} else if pos.Offset > 0 {
			s.offset = int(pos.Offset) + len(string(ch))
		}

		if ch == '\n' {
			s.line = pos.Line + 1
			s.column = 1
		} else {
			s.line = pos.Line
			s.column = pos.Column + uint32(len(string(ch)))
		}
		return ch, pos
	}

	if s.offset >= len(s.source) {
		return 0, Position{}
	}

	pos := Position{
		Offset: uint32(s.offset),
		Line:   s.line,
		Column: s.column,
	}

	ch := rune(s.source[s.offset])
	s.offset++

	if ch == '\r' {
		if s.offset < len(s.source) && rune(s.source[s.offset]) == '\n' {
			s.offset++
		}
		s.line++
		s.column = 1
		return '\n', pos
	}

	if ch == '\n' {
		s.line++
		s.column = 1
		return ch, pos
	}

	s.column++
	return ch, pos
}

func (s *Scanner) peek() rune {
	if len(s.lookahead) > 0 {
		return s.lookahead[0][0].(rune)
	}

	if s.offset >= len(s.source) {
		return 0
	}

	ch := rune(s.source[s.offset])
	if ch == '\r' {
		return '\n'
	}
	return ch
}

func (s *Scanner) peekNth(n int) rune {
	for len(s.lookahead) <= n {
		s.fillLookahead()
	}

	if n < len(s.lookahead) {
		return s.lookahead[n][0].(rune)
	}
	return 0
}

func (s *Scanner) fillLookahead() {
	pos := s.nextLookaheadPosition()
	ch, _ := s.doAdvance()
	if ch != 0 {
		s.lookahead = append(s.lookahead, [2]interface{}{ch, pos})
	}
}

func (s *Scanner) nextLookaheadPosition() Position {
	if len(s.lookahead) > 0 {
		last := s.lookahead[len(s.lookahead)-1]
		ch := last[0].(rune)
		pos := last[1].(Position)
		if ch == '\n' {
			return Position{
				Offset: pos.Offset + 1,
				Line:   pos.Line + 1,
				Column: 1,
			}
		}
		return Position{
			Offset: pos.Offset + uint32(len(string(ch))),
			Line:   pos.Line,
			Column: pos.Column + uint32(len(string(ch))),
		}
	}
	return Position{
		Offset: uint32(s.offset),
		Line:   s.line,
		Column: s.column,
	}
}

func (s *Scanner) unget(ch rune, pos Position) {
	s.lookahead = append([][2]interface{}{{ch, pos}}, s.lookahead...)
	s.offset = int(pos.Offset)
	s.line = pos.Line
	s.column = pos.Column
}

func (s *Scanner) Position() Position {
	return Position{
		Offset: uint32(s.offset),
		Line:   s.line,
		Column: s.column,
	}
}

func (s *Scanner) Offset() uint32 {
	return uint32(s.offset)
}

func (s *Scanner) Line() uint32 {
	return s.line
}

func (s *Scanner) Column() uint32 {
	return s.column
}

func (s *Scanner) IsEof() bool {
	return s.peek() == 0
}

func (s *Scanner) Source() string {
	return s.source
}

func (s *Scanner) Slice(start int, end int) string {
	if start < 0 || end > len(s.source) || start > end {
		return ""
	}
	return s.source[start:end]
}

func (s *Scanner) skipWhile(pred func(rune) bool) {
	for pred(s.peek()) {
		s.advance()
	}
}

func (s *Scanner) consumeIf(ch rune) bool {
	if s.peek() == ch {
		s.advance()
		return true
	}
	return false
}

func (s *Scanner) consumeIfPred(pred func(rune) bool) rune {
	ch := s.peek()
	if pred(ch) {
		s.advance()
		return ch
	}
	return 0
}

func (s *Scanner) IsPuaChar(ch rune) bool {
	return common.IsPUAEncoded(ch)
}
